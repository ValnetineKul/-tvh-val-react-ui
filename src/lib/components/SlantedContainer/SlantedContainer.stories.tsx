import React from 'react';
import { makeStyles } from '../../../themes/core';
import SlantedContainer from './SlantedContainer';
import { createMeta } from '../../story-utils';
import Icon from '../Icon';
import { QuoteLeftSolid } from '../Icon/icons/commercial';
import Typography from '../Typography';
import Surface from '../Surfaces/Surface';

const useStyles = makeStyles()((theme) => ({
  root: {
    marginLeft: theme.layoutSpacing['spacing/400'],
  },
  customContent: {
    border: `2px dashed ${theme.color['border/onSurface100/focus']}`,
    padding: theme.layoutSpacing['spacing/400'],
    width: '220px',
    textAlign: 'center',
  },
}));

type Props = React.ComponentProps<typeof SlantedContainer>;
const Template = (args: Props) => <SlantedContainer {...args} />;

export const Primitive = Template.bind({});
Primitive.args = {
  children: <Icon icon={QuoteLeftSolid} size="sm" />,
};
Primitive.storyName = 'Icon16';

export const IconMd = Template.bind({});
IconMd.args = {
  children: <Icon icon={QuoteLeftSolid} size="md" />,
};
IconMd.storyName = 'Icon24';

export const Label = Template.bind({});
Label.args = {
  children: <Typography>label</Typography>,
};

const CustomTemplate = (args: Props) => {
  const { classes } = useStyles();

  return (
    <SlantedContainer {...args} className={classes.root}>
      <Surface color="100" className={classes.customContent}>
        <Typography>Content placeholder.</Typography>
        <Typography>To be replaced with custom content component.</Typography>
      </Surface>
    </SlantedContainer>
  );
};

export const Custom = CustomTemplate.bind({});
Custom.args = {};

export default createMeta({
  component: SlantedContainer,
  title: 'Components/SlantedContainers/SlantedContainer',

  parameters: {
    docs: {
      transformSource: (src: string) => src.replace(/\[object Object\]/g, 'Typography'),
    },
  },
});
