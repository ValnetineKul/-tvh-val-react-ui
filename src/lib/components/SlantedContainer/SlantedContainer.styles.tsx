import { makeStyles } from '../../../themes/core';

type Props = {
  slantOffset: number;
};

const useStyles = makeStyles<Props>({ name: 'SlantedContainer' })((theme, { slantOffset }) => ({
  root: {
    display: 'flex',
  },
  innerRoot: {
    display: 'flex',
    justifyContent: 'center',
    position: 'relative',
    zIndex: 1,
  },
  skewedStart: {
    '&:before': {
      content: '""',
      height: '100%',
      backgroundColor: 'inherit',
      transform: 'skewX(-10deg)',
      width: '80%',
      zIndex: -1,
      position: 'absolute',
      left: -slantOffset,
    },
  },
  skewedEnd: {
    '&:after': {
      content: '""',
      height: '100%',
      width: '80%',
      zIndex: -1,
      position: 'absolute',
      backgroundColor: 'inherit',
      transform: 'skewX(-10deg)',
      right: -slantOffset,
    },
  },
  paddingNone: {
    padding: theme.layoutSpacing['spacing/000'],
  },
  padding200: {
    padding: theme.layoutSpacing['spacing/200'],
  },
  padding300: {
    padding: theme.layoutSpacing['spacing/300'],
  },
}));

export default useStyles;
