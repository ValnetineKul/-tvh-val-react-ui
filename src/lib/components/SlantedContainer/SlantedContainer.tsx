import type { FC, ReactNode } from 'react';
import React, { useRef, useState, useLayoutEffect } from 'react';
import { capitalize } from '@mui/material/utils';
import type { Surface } from '../../../themes/core';
import useStyles from './SlantedContainer.styles';
import { default as SurfaceComponent } from '../Surfaces/Surface';

export interface SlantedContainerProps {
  className?: string;
  slant?: 'both' | 'start' | 'end';
  surfaceColor?: Surface;
  padding?: 'none' | '200' | '300';
  children: ReactNode;
}

const SlantedContainer: FC<SlantedContainerProps> = ({
  className,
  slant = 'both',
  surfaceColor = 'Primary',
  padding = 'none',
  children,
}) => {
  const ref = useRef<HTMLDivElement>(null);
  const [slantOffset, setSlantOffset] = useState(0);
  const { classes, cx } = useStyles({ slantOffset });
  const paddingClassName = (classes as Record<string, string>)[`padding${capitalize(padding)}`];
  const skewedClassName = cx({
    [classes.skewedStart]: slant === 'both' || slant === 'start',
    [classes.skewedEnd]: slant === 'both' || slant === 'end',
  });

  useLayoutEffect(() => {
    const handleResize = (entries: ResizeObserverEntry[]) => {
      const entry = entries[0];
      setSlantOffset(entry.contentRect.height * 0.09);
    };

    if (!ref.current) {
      return;
    }

    const resizeObserver = new ResizeObserver(handleResize);
    resizeObserver.observe(ref.current);

    return () => {
      resizeObserver.disconnect();
    };
  }, []);

  return (
    <div className={classes.root}>
      <SurfaceComponent color={surfaceColor} className={cx(classes.innerRoot, skewedClassName, className)} ref={ref}>
        <div className={cx(paddingClassName)}>{children}</div>
      </SurfaceComponent>
    </div>
  );
};

export default SlantedContainer;
