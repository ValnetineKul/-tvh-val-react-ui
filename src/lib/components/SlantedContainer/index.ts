import SlantedContainer from './SlantedContainer';

export type { SlantedContainerProps } from './SlantedContainer';
export default SlantedContainer;
