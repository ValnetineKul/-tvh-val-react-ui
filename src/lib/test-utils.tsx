import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
