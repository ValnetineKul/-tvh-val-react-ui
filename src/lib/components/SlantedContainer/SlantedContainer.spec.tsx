import React from 'react';
import type { ComponentProps } from 'react';
import { render, screen } from '../../test-utils';
import SlantedContainer from './SlantedContainer';

type Props = ComponentProps<typeof SlantedContainer>;

describe('SlantedContainer', () => {
  it('Should render correctly', () => {
    const { baseElement } = render(<SlantedContainer>label</SlantedContainer>);
    expect(baseElement).toMatchSnapshot();
  });

  describe('Should set slant', () => {
    const cases: [Props['slant'], string][] = [
      [undefined, 'SlantedContainer-skewedStart-SlantedContainer-skewedEnd'],
      ['start', 'SlantedContainer-skewedStart'],
      ['end', 'SlantedContainer-skewedEnd'],
    ];
    it.each(cases)('given %p, returns %p', (input, expected) => {
      render(<SlantedContainer slant={input}>SlantedContainer</SlantedContainer>);
      const text = screen.getByText('SlantedContainer');
      expect(text?.parentElement?.className).toMatch(expected);
    });
  });

  describe('Should set padding', () => {
    const cases: [Props['padding'], string][] = [
      [undefined, 'SlantedContainer-paddingNone'],
      ['200', 'SlantedContainer-padding200'],
      ['300', 'SlantedContainer-padding300'],
    ];
    it.each(cases)('given %p, returns %p', (input, expected) => {
      render(<SlantedContainer padding={input}>SlantedContainer</SlantedContainer>);
      const text = screen.getByText('SlantedContainer');
      expect(text.className).toMatch(expected);
    });
  });
});
