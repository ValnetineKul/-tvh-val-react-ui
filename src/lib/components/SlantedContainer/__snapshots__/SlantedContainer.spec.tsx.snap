// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlantedContainer Should render correctly 1`] = `
<body>
  <div>
    <div
      class="tss-8mr7tx-SlantedContainer-root"
    >
      <div
        class="tss-1j76svm-Surface-root-SlantedContainer-innerRoot-SlantedContainer-skewedStart-SlantedContainer-skewedEnd"
      >
        <div
          class="tss-1w4qckr-SlantedContainer-paddingNone"
        >
          label
        </div>
      </div>
    </div>
  </div>
</body>
`;
